sdk/lib/queue/queue.cc

```diff
@@ -0,0 +1,435 @@
+#include <cheri.hh>
+#include <errno.h>
+#include <locks.hh>
+#include <queue.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <timeout.h>
+#include <type_traits>
+
+using namespace CHERI;
+using cheriot::atomic;
+
+using Debug = ConditionalDebug<false, "Queue library">;
+
+#ifdef __cplusplus
+using MessageQueueCounter = atomic<uint32_t>;
+#else
+typedef uint32_t MessageQueueCounter;
+#endif
+
+#include <assembly-helpers.h>
+
+namespace
+{
+	/**
+	 * Helpers for the queue.  The queue uses two counters that wrap on double
+	 * the number of elements.  This ensures that full and empty conditions
+	 * have different values: a queue is full when the producer is one queue
+	 * length ahead of the consumer, and empty when the producer and consumer
+	 * are equal.
+	 */
+
+	/**
+	 * Helper for wrapping increment.  Increments `counter`, wrapping to zero
+	 * if it reaches double `size`.
+	 */
+	constexpr uint32_t increment_and_wrap(uint32_t size, uint32_t counter)
+	{
+		counter++;
+		if (2 * size == counter)
+		{
+			return 0;
+		}
+		return counter;
+	}
+
+	/**
+	 * Returns true if and only if the `size`, `producer`, and `consumer`
+	 * counters indicate a full queue.
+	 */
+	constexpr bool is_full(uint32_t size, uint32_t producer, uint32_t consumer)
+	{
+		return producer - consumer == size;
```

Comment:
Shouldn't this check for `producer < consumer`?

Reply:
The logic was also wrong in the items-remaining check.  I have now added some exciting recursive template magic to do compile-time exhaustive checking.